add ID as default filter if no other filtering criteria is provided

rebeccahhh opened the pull request:
have instances be filtered by ID in case of no filtering criteria passed in per the sorting issues present in https://github.com/ansible/awx/issues/11740

<!--- changelog-entry
# Fill in 'msg' below to have an entry automatically added to the next release changelog.
# Leaving 'msg' blank will not generate a changelog entry for this PR.
# Please ensure this is a simple (and readable) one-line string.
---
msg: "add ID as default filter if no other filtering criteria is provided"
-->

##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->

<!---
If you are fixing an existing issue, please include "related #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Bugfix Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
 - API


##### AWX VERSION
<!--- Paste verbatim output from `make VERSION` between quotes below -->
```
awx: 20.0.2
```

rebeccahhh committed:
have instances be filtered by ID in case of no filtering criteria passed in

and then switch from using order by ID as a fallback for all ordering and instead
just set instances ordering to ID as default to prevent
OrderedManyToMany fields ordering from being interrupted.

shanemcd commented:
@rebeccahhh Are you looking into the test failures here?

rebeccahhh commented:
@shanemcd yes, changes incoming to prevent other things from being affected.

AlanCoding commented on awx/api/filters.py:
```diff
@@ -395,11 +395,11 @@ def filter_queryset(self, request, queryset, view):
-                order_by = self.get_default_ordering(view)
+            default_order_by = self.get_default_ordering(view)
+            # glue the order by and default order by together so that the default is the backup option
+            order_by = list(order_by or []) + list(default_order_by or [])
```
With the change to the view ordering, I'm not sure if this change is necessary.  I also don't like that it would be changing the filter class behavior in the case that `order_by` is `None`.

jbradberry replied:
This change is absolutely necessary.  The original problem is when someone has clicked on the 'status' header, the sort is unstable.  This is because the usual case is that the value is the empty string.  Our goal here is to have the parameters passed to Django's `.order_by()` be `['errors', 'id']`, so that there is a fall-back field resulting in a stable sort.

jbradberry replied:
Previously when the UI passes in _any_ parameter for `order_by`, the default ordering specified by the view gets thrown away entirely.  Now it will be appended to the end of any parameters passed in by the UI as a fall back.

jbradberry replied:
There _may_ be instances where this changes the actual ordering that a view results in, but it can only come up when a user-requested sort is used and _only_ as the final fall-back field, so I think it will be quite rare and unobtrusive.

AlanCoding replied:
This is going to impact _all_ list views. I see what you're saying - the prior behavior was that when `order_by` was None it used the view's default ordering (seems unnecessary, but not something this is changing). My last comment was wrong.

But I would like to see a test or two, because this could be a relatively high-impact change. While the issue filed was for instances, I'd expect that similar bugs can be seen in other parts of the UI (order projects by scm_type, right?)

jbradberry replied:
> But I would like to see a test or two, because this could be a relatively high-impact change. While the issue filed was for instances, I'd expect that similar bugs can be seen in other parts of the UI (order projects by scm_type, right?)

Probably.

rebeccahhh replied:
I am currently writing some tests. If there are any particular scenarios you're concerned about let me know so I can account for them specifically.

AlanCoding replied:
@rebeccahhh Create a decent number (more than 5) projects with randomly generated names. Use the test fixtures to simulate a GET to the projects list endpoint, and loop over the results, assuring that they have the default ordering applied, which is by id, which comes from the model definition.

rebeccahhh committed:
add projects to test that ordering functions correctly and when it gets a value it cannot order by it falls back to ID

add tests that check ordering for projects, organizations, inventories, groups, and hosts

CFSNM commented:
Created PR in tower-qa to add some automated test to cover this issue: https://github.com/ansible/tower-qa/pull/7346

AlanCoding commented on awx/api/views/__init__.py:
```diff
@@ -364,6 +364,7 @@ class InstanceList(ListAPIView):
     model = models.Instance
     serializer_class = serializers.InstanceSerializer
     search_fields = ('hostname',)
+    ordering = ('id',)
```
The default ordering of the model is by hostname

https://github.com/ansible/awx/blob/56f51eebce7004630394c1c76a15eaebd652072e/awx/main/models/ha.py#L142

I don't see any reason to introduce this ordering by id.

sarabrajsingh replied:
doesn't this need to be `id` so that the sorting works with other drop-downs/views?

jbradberry replied:
@AlanCoding the issue is when the user overrides sorting by clicking on the table headers.  The default model ordering is thrown out in that case, and the reason this has been brought up is for situations where the user is trying to sort by something that has a lot of indistinguishable values, e.g. the errors field.  We absolutely need this view default ordering, because that is what we are using in the changes in this PR to build up a multi-field set of sorting criteria.  An argument could be made that maybe we should try to do something clever and fold in _all_ sources of default ordering, but that would have been even more complex and potentially fragile.

CFSNM commented:
Hi, I left this comment in the QE PR. Can anybody take a look?
https://github.com/ansible/tower-qa/pull/7346#issuecomment-1083263071

CFSNM reviewed (approved):
Merged PR https://github.com/ansible/tower-qa/pull/7346 to add QE coverage. OK to merge this one

rebeccahhh commented:
we already have testing for ordering surrounding instances and instance groups, I don't think there's much more to cover here, anyone disagreeing ping me so I can add more testing to cover it (also interactive rebased so that there are only 2 commits, one for the bugfix and one for the tests).

rebeccahhh commented on awx/main/tests/functional/test_projects.py:
```diff
@@ -408,3 +408,113 @@ def test_project_delete(delete, organization, admin_user):
+    assert project_names == expected_names and project_ids == expected_ids
+
+
+@pytest.mark.parametrize('order_by', ('name', '-name'))
```
@AlanCoding I think the below test covers what you were hoping for it to cover.

rebeccahhh committed:
I don't think these tests actually add anything, so I am removing them even though I wrote them in the first place.

kialam commented on awx/main/tests/functional/test_projects.py:
```diff
@@ -408,3 +408,46 @@ def test_project_delete(delete, organization, admin_user):
+
+@pytest.mark.parametrize('order_by', ('name', '-name'))
+@pytest.mark.django_db
+def test_project_list_ordering_with_duplicate_names(get, order_by, organization_factory):
```
👍

kialam commented on awx/main/tests/functional/test_projects.py:
```diff
@@ -408,3 +408,46 @@ def test_project_delete(delete, organization, admin_user):
+)
+@pytest.mark.django_db
+def test_project_list_ordering_by_name(get, order_by, expected_names, expected_ids, organization_factory):
+    'ensure sorted order of project list is maintained correctly when the requested order is invalid or not applicable'
```
👍
